recipes/brpc/all/conanfile.py

```diff
@@ -0,0 +1,98 @@
+from conans import CMake, ConanFile, tools
+import glob
+import os
+
+
+class BrpcConan(ConanFile):
+    name = "brpc"
+    description = "An industrial-grade RPC framework used throughout Baidu"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/apache/incubator-brpc"
+    topics = ("conan", "brpc", "baidu", "rpc")
+    license = ("Apache-2.0")
+    exports_sources = ["CMakeLists.txt", "patches/**"]
+    generators = "cmake", "cmake_find_package"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "with_snappy": [True, False],
+        "with_glog": [True, False],
+        "with_thrift": [True, False],
+        "shared": [True, False],
+        "fPIC": [True, False]
+    }
+    default_options = {
+        "with_snappy": True,
+        "with_glog": True,
+        "with_thrift": False,
+        "shared": False,
+        "fPIC": True,
+    }
+
+    _cmake = None
+
+    def requirements(self):
+        self.requires("gflags/2.2.2")
+        self.requires("protobuf/3.9.1")
+        self.requires("leveldb/1.22")
+        self.requires("openssl/1.1.1g")
+        self.requires("zlib/1.2.11")
+        if self.options.with_glog:
+            self.requires("glog/0.4.0")
+        if self.options.with_thrift:
+            self.requires("thrift/0.13.0")
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def configure(self):
+        self.options['protobuf'].with_zlib = True
+        self.options['leveldb'].with_snappy = self.options.with_snappy
```

Comment:
The general rule is to never modify options of dependencies.
Also, options of dependencies are not initialized in the `configure` method.
You must check for options of dependencies in the `build` method.
There you should do:
```
        if not self.options["protobuf"].with_zlib:
            raise ConanInvalidConfiguration("brpc requires protobuf:with_zlib=True")
        if self.options.with_snappy and not self.options["leveldb"].with_snappy:
            raise ConanInvalidConfiguration("brpc requires leveldb:with_snappy=True")
```

Reply:
Thanks for the feedback. I removed the explicit assignment of these options and added checks to the build method as you suggested. TBH I was very unclear on `config_options` vs `configure` so I read through a couple of docs. https://docs.conan.io/en/latest/mastering/conditional.html states: `The configure method might be used to hardcode dependencies options values. ... Use it only when it is absolutely necessary that the package dependencies use those options.` What is still unclear is what meets the bar of "absolutely necessary". I guess the documentation would benefit from an example to clarify when this might be justified.

Anyway, I took your advice, as well as the advice in that page (as I understood it) and added "default options" for `protobuf:with_zlib` and `leveldb:with_snappy`. Thanks for your time, and let me know if the changes look reasonable.

Reply:
It's *legal* for a conan recipe to hardcode options of dependencies,
but the recipes at cci try to do no magic in order to not confuse people.
Our belief it's better to inform people then to hard code potential hard-to-debug behavior.
Maybe this is a point where conan/cci requires more polish.